### ISSUE TITLE
Unsupported methods on REST endpoints should respond with status code 405

### ISSUE THREAD

**Issue by jasontedor** [elastic/elasticsearch#15335]
Today, if a request is executed against an endpoint with a method in the request line that that endpoint does not support, Elasticsearch responds with a response like:

```
{
  "error" : {
    "root_cause" : [ {
      "type" : "illegal_argument_exception",
      "reason" : "No feature for name [_forcemerge]"
    } ],
    "type" : "illegal_argument_exception",
    "reason" : "No feature for name [_forcemerge]"
  },
  "status" : 400
}
```

This is confusing for users*, semantically wrong, and contrary to the HTTP spec. Instead, Elasticsearch should respond with [HTTP status code 405](https://tools.ietf.org/html/rfc2616#section-10.4.6) (Method Not Allowed) and include in the Allow header a list of methods that the endpoint does support.

*: [Googling "no feature for name elasticsearch"](https://www.google.com/search?q=no+feature+for+name+elasticsearch) just leads to pages describing issues when an [index is missing](https://github.com/elastic/elasticsearch/blob/20bff773eda14804a77f6b76d98d8ea63ba6f4b9/core/src/main/java/org/elasticsearch/action/admin/indices/get/GetIndexRequest.java#L75-L82).

**Comment by dakrone** [comment]
+1

**Comment by nik9000** [comment]
+1

**Comment by jbertouch** [comment]
@jasontedor Made a start on this, but it still needs unit tests. Are there existing http endpoint exception/http status code tests where new tests could be included? Had a quick look and couldn't find an appropriate home for them.

Also noticed that the `RestController#executeHandler` didn't have a complete implementation of the [http OPTIONS method](https://tools.ietf.org/html/rfc2616#page-52), so I cleaned it up. The prior code just returned 200 OK for all requests, and didn't included allowed methods for the endpoint as recommended in the spec.

**Comment by dakrone** [comment]
> Are there existing http endpoint exception/http status code tests where new tests could be included? Had a quick look and couldn't find an appropriate home for them.

Usually this would go into the YAML REST tests, but for a special case like this you need a bit more powerful matching than that provides, so probably a test subclassing `ESSingleNodeTestCase` that uses `java.net.URLConnection` would work?

**Comment by jbertouch** [comment]
Thanks @dakrone

**Comment by olcbean** [comment]
I cannot reproduce this in `7.0` 
```
GET _forcemerge
```
```
{
  "error": "Incorrect HTTP method for uri [/_forcemerge] and method [GET], allowed: [POST]",
  "status": 405
}
```
I think this has been fixed by #24437.
cc @dakrone @jasontedor

**Comment by dakrone** [comment]
Yep that's right, this can be closed now, thanks @olcbean
